### PR TITLE
[Merged by Bors] - Modularise slasher backend

### PR DESCRIPTION
**Pull request by michaelsproul**
## Proposed Changes

Enable multiple database backends for the slasher, either MDBX (default) or LMDB. The backend can be selected using `--slasher-backend={lmdb,mdbx}`.

## Additional Info

In order to abstract over the two library's different handling of database lifetimes I've used `Box::leak` to give the `Environment` type a `'static` lifetime. This was the only way I could think of using 100% safe code to construct a self-referential struct `SlasherDB`, where the `OpenDatabases` refers to the `Environment`. I think this is OK, as the `Environment` is expected to live for the life of the program, and both database engines leave the database in a consistent state after each write. The memory claimed for memory-mapping will be freed by the OS and appropriately flushed regardless of whether the `Environment` is actually dropped.

We are depending on two `sigp` forks of `libmdbx-rs` and `lmdb-rs`, to give us greater control over MDBX OS support and LMDB's version.

### CODE REVIEW

**Comment by michaelsproul** [comment]
TODO

- [x] Re-open https://github.com/sigp/lighthouse/issues/2342, which still isn't fixed in an LMDB release.
- [x] Run slasher tests with LMDB on CI

**Comment by michaelsproul** [comment]
I've marked this ready for review. I haven't updated the LMDB version because I don't have time to test on Windows right now. I figure we can come back for that in a subsequent update. Windows users can use MDBX in the meantime.

File: beacon_node/src/cli.rs [hunk 1]
```diff
@@ -622,6 +623,14 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
                        [disabled by default].")
                 .requires("slasher")
         )
+        .arg(
+            Arg::with_name("slasher-backend")
+                .long("slasher-backend")
+                .help("Set the database backend to be used by the slasher.")
+                .takes_value(true)
+                .possible_values(slasher::DatabaseBackend::VARIANTS)
```
Comment (paulhauner) [inline beacon_node/src/cli.rs:631]: This is cool!

**Review by paulhauner — APPROVED** [review]
I've given this a read-through and it looks good! I agree with targeting this for v3.0.0.

File: beacon_node/src/cli.rs (hunk 1 above)
Comment (michaelsproul) [reply beacon_node/src/cli.rs:631]: since I worked this out I've been putting it in all my PRs haha

**Comment by michaelsproul** [comment]
I've been running both backends on mainnet for 5 days (since Aug 10 00:00 UTC) and the current stats are as follows:

- MDBX
  + Database size: 22 GiB
  + Current time per batch: 1.3s
- LMDB
  + Database size: 27 GiB 
  + Current time per batch: 1.2s

So LMDB is definitely still viable (and even a bit faster!). At this rate MDBX will use around 80 GiB for the full history of 4096 epochs and LMDB will use 100 GiB, around 25% more. The real test will begin once both DBs have surpassed 4096 epochs of history and MDBX's superior pruning kicks in.

**Comment by paulhauner** [comment]
Pentasquerge! :rocket: :rocket: :rocket: :rocket: :rocket:

bors r+

**Pull request closed without merging**